### PR TITLE
Implement Mutable FST Index

### PR DESCRIPTION
**Pull request by atris**
This PR implements a real time FST index, allowing fields containing tokenized documents to be directly stored in the FST index and queried in real time, with no flush requirement.

This can be enabled by specifying the following property in the properties section of the FST index, being created on a real time table:
`
 "properties":[{"fstType":"native"}]`

### CODE REVIEW

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/realtime/impl/invertedindex/NativeMutableFSTIndex.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+import org.apache.pinot.segment.spi.index.mutable.MutableTextIndex;
+import org.roaringbitmap.RoaringBitmapWriter;
+import org.roaringbitmap.buffer.ImmutableRoaringBitmap;
+import org.roaringbitmap.buffer.MutableRoaringBitmap;
+
+
+public class NativeMutableFSTIndex implements MutableTextIndex {
+  private final String _column;
+  private final MutableFST _fst;
+  private int _dictId;
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/realtime/impl/invertedindex/NativeMutableFSTIndex.java:35]: (minor)
Move this after the final variables for readability

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/realtime/impl/invertedindex/NativeMutableFSTIndex.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+      RealTimeRegexpMatcher.regexMatch(searchQuery, _fst, writer::add);
+      return writer.get();
+    } finally {
+      _readLock.unlock();
+    }
+  }
+
+  @Override
+  public MutableRoaringBitmap getDocIds(String searchQuery) {
+    return null;
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/realtime/impl/invertedindex/NativeMutableFSTIndex.java:73]: Let's throw an unsupported exception with some error message for debugging purpose

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/indexsegment/mutable/MutableSegmentImpl.java [hunk 3]
```diff
@@ -306,6 +307,21 @@ public long getLatestIngestionTimestamp() {
       MutableInvertedIndex invertedIndexReader =
           invertedIndexColumns.contains(column) ? indexProvider.newInvertedIndex(context.forInvertedIndex()) : null;
 
+      MutableTextIndex fstIndex = null;
+      //FST Index
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/indexsegment/mutable/MutableSegmentImpl.java:311]: (nit) Add a space after the `//`

File: pinot-segment-local/src/test/java/org/apache/pinot/segment/local/realtime/impl/invertedindex/NativeMutableFSTIndexTest.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    return Arrays.asList("Prince", "Andrew", "kept", "looking", "with", "an", "amused", "smile", "from", "Pierre");
+  }
+
+  private void testSelectionResults(String nativeQuery, int resultCount, @Nullable List<Integer> results) {
+    ImmutableRoaringBitmap resultMap = _nativeMutableFSTIndex.getDictIds(nativeQuery);
+    assertEquals(resultMap.getCardinality(), resultCount);
+
+    if (results != null) {
+      for (int result : results) {
+        assertEquals(resultMap.contains(result), true);
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/test/java/org/apache/pinot/segment/local/realtime/impl/invertedindex/NativeMutableFSTIndexTest.java:79]: (minor) `assertTrue`

**Review by Jackie-Jiang — APPROVED** [review]
Good job! Only minor comments

Please add some description to the PR, and add a release-note section

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/indexsegment/mutable/MutableSegmentImpl.java [hunk 5]
```diff
@@ -306,6 +307,21 @@ public long getLatestIngestionTimestamp() {
...
+      if (fstIndexColumns.contains(column)) {
+        if (_fieldConfigList != null) {
```
Comment (jackjlli) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/indexsegment/mutable/MutableSegmentImpl.java:313]: nit: you can combine these nested if statement into one like:
 `if (_fieldConfigList != null && fstIndexColumns.contains(column))`

File: pinot-segment-local/src/test/java/org/apache/pinot/segment/local/realtime/impl/invertedindex/NativeMutableFSTIndexTest.java [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    List<Integer> resultList = Arrays.asList(0, 9);
+    testSelectionResults(nativeQuery, 2, resultList);
+
+    nativeQuery = "a.*";
+    resultList = Arrays.asList(5, 6);
+    testSelectionResults(nativeQuery, 2, resultList);
+
+    nativeQuery = ".*ed";
+    resultList = Arrays.asList(6);
+    testSelectionResults(nativeQuery, 1, resultList);
```
Comment (jackjlli) [inline pinot-segment-local/src/test/java/org/apache/pinot/segment/local/realtime/impl/invertedindex/NativeMutableFSTIndexTest.java:66]: Add more tests like `.*m.*`

**Pull request merged**